Mathlib/NumberTheory/NumberField/Units.lean

```diff
@@ -149,4 +165,416 @@ theorem rootsOfUnity_eq_torsion [NumberField K] :
+
+theorem logEmbedding_component_le {r : ℝ} {x : (𝓞 K)ˣ} (hr : 0 ≤ r) (h : ‖logEmbedding K x‖ ≤ r)
+    (w : {w : InfinitePlace K // w ≠ w₀}) : |logEmbedding K x w| ≤ r := by
+  lift r to NNReal using hr
+  simp_rw [Pi.norm_def, NNReal.coe_le_coe, Finset.sup_le_iff, ← NNReal.coe_le_coe] at h
+  exact h w (mem_univ _)
+
+theorem log_le_of_logEmbedding_le {r : ℝ} {x : (𝓞 K)ˣ} (hr : 0 ≤ r) (h : ‖logEmbedding K x‖ ≤ r)
+    (w : InfinitePlace K) : |Real.log (w x)| ≤ (Fintype.card (InfinitePlace K)) * r := by
+  have tool : ∀ x : ℝ, 0 ≤ x → x ≤ mult w * x := fun x hx => by
+      nth_rw 1 [← one_mul x]
+      refine mul_le_mul ?_ le_rfl hx ?_
+      all_goals { rw [mult]; split_ifs <;> norm_num }
+  by_cases hw : w = w₀
+  · have hyp := congr_arg (‖·‖) (sum_logEmbedding_component x).symm
+    replace hyp := (le_of_eq hyp).trans (norm_sum_le _ _)
+    simp_rw [norm_mul, norm_neg, Real.norm_eq_abs, Nat.abs_cast] at hyp
+    refine (le_trans ?_ hyp).trans ?_
+    · rw [← hw]
+      exact tool _ (abs_nonneg _)
+    · refine (sum_le_card_nsmul univ _  _
+        (fun w _ => logEmbedding_component_le hr h w)).trans ?_
+      rw [nsmul_eq_mul]
+      refine mul_le_mul ?_ le_rfl hr (Fintype.card (InfinitePlace K)).cast_nonneg
+      simp [card_univ]
+  · have hyp := logEmbedding_component_le hr h ⟨w, hw⟩
+    rw [logEmbedding_component, abs_mul, Nat.abs_cast] at hyp
+    refine (le_trans ?_ hyp).trans ?_
+    · exact tool _ (abs_nonneg _)
+    · nth_rw 1 [← one_mul r]
+      exact mul_le_mul (Nat.one_le_cast.mpr Fintype.card_pos) (le_of_eq rfl) hr (Nat.cast_nonneg _)
+
+variable (K)
+
+/-- The lattice formed by the image of the logarithmic embedding. -/
+noncomputable def _root_.NumberField.Units.unitLattice :
+    AddSubgroup ({w : InfinitePlace K // w ≠ w₀} → ℝ) :=
+  AddSubgroup.map (logEmbedding K) ⊤
+
+theorem unitLattice_inter_ball_finite (r : ℝ) :
+    ((unitLattice K : Set ({ w : InfinitePlace K // w ≠ w₀} → ℝ)) ∩
+      Metric.closedBall 0 r).Finite := by
+  obtain hr | hr := lt_or_le r 0
+  · convert Set.finite_empty
+    rw [Metric.closedBall_eq_empty.mpr hr]
+    exact Set.inter_empty _
+  · suffices {x : (𝓞 K)ˣ | IsIntegral ℤ (x : K) ∧
+        ∀ (φ : K →+* ℂ), ‖φ x‖ ≤ Real.exp ((Fintype.card (InfinitePlace K)) * r)}.Finite by
+      refine (Set.Finite.image (logEmbedding K) this).subset ?_
+      rintro _ ⟨⟨x, ⟨_, rfl⟩⟩, hx⟩
+      refine ⟨x, ⟨x.val.prop, (le_iff_le _ _).mp (fun w => (Real.log_le_iff_le_exp ?_).mp ?_)⟩, rfl⟩
+      · exact pos_iff.mpr (coe_ne_zero x)
+      · rw [mem_closedBall_zero_iff] at hx
+        exact (le_abs_self _).trans (log_le_of_logEmbedding_le hr hx w)
+    refine Set.Finite.of_finite_image ?_ ((coe_injective K).injOn _)
+    refine (Embeddings.finite_of_norm_le K ℂ
+        (Real.exp ((Fintype.card (InfinitePlace K)) * r))).subset ?_
+    rintro _ ⟨x, ⟨⟨h_int, h_le⟩, rfl⟩⟩
+    exact ⟨h_int, h_le⟩
+
+section span_top
+
+/-!
+## Section `span_top`
```

Comment:
```suggestion
### Section `span_top`
```
Not 100% sure...

Reply:
Actually, I think I want `#### Section span_top` so that it looks like a subsection of `### Dirichlet Unit Theorem` if my understanding of the existing code is correct.

Reply:
I have no idea if the number of `#` is unlimited, but we can try and at worst change this in a small PR later.